### PR TITLE
Solution to OOP Advanced

### PR DESCRIPTION
**Pull request by nut-pine**

### CODE REVIEW

File: src/main/java/core/basesyntax/Circle.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class Circle extends Shape {
+
+    private Random rand;
+    private StringBuilder sb = new StringBuilder();
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:8]: sb - it's informative naming, let's improve this

File: src/main/java/core/basesyntax/Circle.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    private double roundness; // Окружность
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:9]: eng please

File: src/main/java/core/basesyntax/Circle.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package core.basesyntax;
+
+import java.util.Random;
+
+public class Circle extends Shape {
+
+    private Random rand;
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:7]:
```suggestion
    private Random random;
```

File: src/main/java/core/basesyntax/Circle.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    public Circle(String name) {
+        super(name);
+        rand = new Random();
+        this.roundness = rand.nextInt(50);
+    }
+
+    @Override
+    public double calculateArea() {
+        return (int) (radius() * Math.PI);
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:19]: why cast to int?

File: src/main/java/core/basesyntax/IsoscelesTrapezoid.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    private Random rand;
+    private StringBuilder sb = new StringBuilder();
+    private int sideA;
+    private int sideB;
+    private int sideC;
+
+    public IsoscelesTrapezoid(String name) {
+        super(name);
+        rand = new Random();
+        this.sideA = rand.nextInt(10);
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/IsoscelesTrapezoid.java:16]: it's not clear why do we use random parameters to init the object. 
In the constructor take ready to use parameters and create a new object based on that. For creating random figures use a separate class FigureProducer

File: src/main/java/core/basesyntax/Main.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        for (Shape s : shapeArr) {
+            System.out.println(s.info());
+        }
+    }
+
+    public Shape[] fill() {
+        Random rand = new Random();
+        Shape[] shapeArr = new Shape[generateRandomArrayLength()];
+        Shape shape = null;
+        for (int i = 0; i < shapeArr.length; i++) {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Main.java:20]: this logic better to have in the separate class

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Circle.java (hunk 1 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: fixed!

File: src/main/java/core/basesyntax/Circle.java (hunk 2 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: fixed!

File: src/main/java/core/basesyntax/Circle.java (hunk 3 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: fixed!

File: src/main/java/core/basesyntax/Circle.java (hunk 4 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java:15]: I wanted to round a digit. Fixed it. Now I use a method Math.round();

File: src/main/java/core/basesyntax/IsoscelesTrapezoid.java (hunk 5 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/IsoscelesTrapezoid.java]: done!

File: src/main/java/core/basesyntax/Main.java (hunk 6 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Main.java]: done!

File: src/main/java/core/basesyntax/Circle.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    private double roundness;
+
+    public Circle(String name, FigureProducer producer) {
+        super(name);
+        this.roundness = producer.generatorNumber();
+    }
+
+    @Override
+    public double calculateArea() {
+        return (int) (radius() * Math.PI);
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:15]: you still cast it to int, so 3.9 will return 3 which is not that good, let's return double, that will be the optimal solution here

File: src/main/java/core/basesyntax/Circle.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        return radius() * 2 * Math.PI;
+    }
+
+    @Override
+    public void draw() {
+        System.out.print("Drawing Circle");
+    }
+
+    @Override
+    public String info() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:29]:
```suggestion
    public String displayInfo() {
```
methods are mostly verbs

File: src/main/java/core/basesyntax/Circle.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        return message.append(super.info()).append(",area: " + this.calculateArea() + " sq.unit,")
+        .append("radius:" + this.radius() + " units,").toString();
+    }
+
+    public double radius() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:34]:
```suggestion
    public double getRadius() {
```

File: src/main/java/core/basesyntax/Circle.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        this.roundness = producer.generatorNumber();
+    }
+
+    @Override
+    public double calculateArea() {
+        return (int) (radius() * Math.PI);
+    }
+
+    @Override
+    public double perimeter() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Circle.java:19]: same here

File: src/main/java/core/basesyntax/FigureProducer.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        return shapeArr;
+    }
+
+    public static int generateRandomArrayLength() {
+        int generator = (int)(20 + Math.random() * 60);
+        return generator;
+    }
+
+    public int generatorNumber() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/FigureProducer.java:40]:
```suggestion
    public int generateRandomNumber() {
```

File: src/main/java/core/basesyntax/Main.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package core.basesyntax;
+
+public class Main {
+
+    public static void main(String[] args) {
+        Shape[] shapeArr = new FigureProducer().fill();
+
+        for (Shape s : shapeArr) {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Main.java:8]:
```suggestion
        for (Shape shape/element : shapeArr) {
```

File: src/main/java/core/basesyntax/Shape.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    public abstract double perimeter();
+
+    public abstract void draw();
+
+    public String info() {
+        return sb.append("Shape: ").append(this.getName())
+        .append(", color: " + this.getColor()).toString();
+    }
+
+    private String randomColor() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Shape.java:33]: move this logic to your FIgureProducer or create separate class

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Circle.java (hunk 7 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: done!

File: src/main/java/core/basesyntax/Circle.java (hunk 8 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: done!

File: src/main/java/core/basesyntax/Circle.java (hunk 10 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: done!

File: src/main/java/core/basesyntax/FigureProducer.java (hunk 11 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/FigureProducer.java]: done!

File: src/main/java/core/basesyntax/Main.java (hunk 12 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Main.java]: done!

File: src/main/java/core/basesyntax/Shape.java (hunk 13 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Shape.java]: done!

File: src/main/java/core/basesyntax/Circle.java (hunk 9 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Circle.java]: done!

File: src/main/java/core/basesyntax/Square.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public String displayInfo() {
+        return message.append(super.displayInfo())
+        .append(",area: " + this.calculateArea() + " sq.units ")
+        .append("side length: " + this.sideLength() + " units ").toString();
+
+    }
+
+    public double sideLength() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Square.java:35]:
```suggestion
    public double getSideLength() {
```
point about verbs is valid for all mehtods

File: src/main/java/core/basesyntax/Shape.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+    public String getColor() {
+        return color;
+    }
+
+    public abstract double calculateArea();
+
+    public abstract double getPerimeter();
+
+    public abstract void draw();
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Shape.java:26]: these methods should be in the inteface

File: src/main/java/core/basesyntax/Shape.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+    public String displayInfo() {
+        return message.append("Shape: ").append(this.getName())
+        .append(", color: " + this.getColor()).toString();
+    }
+}
+
+
+
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/Shape.java:35]: remove redundant empty lines

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]
move methods from class to interface, remove redundant empty lines and you missed one method while naming fix, rest is good

File: src/main/java/core/basesyntax/Square.java (hunk 14 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Square.java]: done!

File: src/main/java/core/basesyntax/Shape.java (hunk 15 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Shape.java]: done!

File: src/main/java/core/basesyntax/Shape.java (hunk 16 above)
Comment (nut-pine) [reply src/main/java/core/basesyntax/Shape.java]: done!

File: src/main/java/core/basesyntax/ColorProducer.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package core.basesyntax;
+
+public class ColorProducer {
+
+    public String generateRandomNumber() {
```
Comment (RomanDubovskyi) [inline src/main/java/core/basesyntax/ColorProducer.java:5]:
```suggestion
    public String generateRandomColor() {
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request closed without merging**